api/apps/geoprocessing/src/modules/planning-units-grid/planning-units-grid.processor.ts

```diff
@@ -0,0 +1,76 @@
+import { Injectable, Logger } from '@nestjs/common';
+import { Job, Worker } from 'bullmq';
+import { plainToClass } from 'class-transformer';
+import { validateSync } from 'class-validator';
+import { InjectEntityManager } from '@nestjs/typeorm';
+import { EntityManager } from 'typeorm';
+
+import { JobInput, JobOutput, queueName } from '@marxan/planning-units-grid';
+import { ShapefileService } from '@marxan/shapefile-converter';
+import {
+  WorkerBuilder,
+  WorkerProcessor,
+} from '@marxan-geoprocessing/modules/worker';
+
+import { ShapeType } from '@marxan-jobs/planning-unit-geometry';
+import { GridGeoJsonValidator } from './grid-geojson-validator';
+import { isLeft } from 'fp-ts/Either';
+
+@Injectable()
+export class PlanningUnitsGridProcessor
+  implements WorkerProcessor<JobInput, JobOutput> {
+  private readonly worker: Worker;
+  private readonly logger = new Logger(this.constructor.name);
+
+  constructor(
+    @InjectEntityManager() private readonly entityManager: EntityManager,
+    private readonly shapefileConverter: ShapefileService,
+    private readonly gridGeoJsonValidator: GridGeoJsonValidator,
+    workerBuilder: WorkerBuilder,
+  ) {
+    this.worker = workerBuilder.build(queueName, this);
+  }
+
+  async process({
+    data: { projectId, shapefile },
+  }: Job<JobInput, JobOutput>): Promise<JobOutput> {
+    const { data: geoJson } = await this.shapefileConverter.transformToGeoJson(
+      shapefile,
+    );
+
+    const result = this.gridGeoJsonValidator.validate(geoJson);
+    if (isLeft(result)) {
+      throw new Error(result.left.toString());
+    }
+    const puGeometriesIds: { id: string }[] = await this.entityManager.query(
+      `
+        INSERT INTO "planning_units_geom"("the_geom", "type", "project_id")
+        SELECT ST_SetSRID(
+                 ST_GeomFromGeoJSON(features ->> 'geometry'),
+                 4326)::geometry,
+               $2::shape_type,
+               $3
+        FROM (
+               SELECT json_array_elements($1::json -> 'features') AS features
+             ) AS f
+        ON CONFLICT ON CONSTRAINT planning_units_geom_the_geom_type_key DO UPDATE SET type = 'from_shapefile'::shape_type
```

Comment:
I am not too sure about this - I guess you and Alicia discussed this so it will be ok - at a glance, I'm not sure this caters to a desired use case, but I also see no harm as we're updating existing shapes within the boundaries of a project.

Reply:
@hotzevzl the only reason I update is I need to, otherwise IDs won't be returned back. Seems like there is no harm (could as well update projectId)